internal/controller/externalip_controller.go

```diff
@@ -78,290 +75,118 @@ func (r *ExternalIPReconciler) Reconcile(ctx context.Context, req ctrl.Request)
-	}
-
-	// 2nd STEP
-	//
-	// Release unassociated address.
-	if externalIP.IsReserved() {
-		// Do not delete EIP if flag PreventEIPDeallocation is set
-		if externalIP.Status.AddressID != nil && !externalIP.Spec.PreventEIPDeallocation {
-			if err := r.Provider.DeleteAddress(ctx, *externalIP.Status.AddressID); err != nil {
-				if !provider.IsErrNotFound(err) {
-					log.Error(err, "Failed to delete Address", "addressID", *externalIP.Status.AddressID)
-					return ctrl.Result{}, err
-				}
-				log.V(1).Info("Address not found", "addressID", *externalIP.Status.AddressID)
-			}
-			log.Info("Deleted Address", "addressID", *externalIP.Status.AddressID)
-
-			// Update status
-			externalIP.Status.AddressID = nil
-		}
-		// set State to None for finalizer to delete externalIP object
-		externalIP.Status.State = v1alpha1.ExternalIPStateNone
-		log.V(1).Info("Updating ExternalIP", "state", externalIP.Status.State)
-		return ctrl.Result{RequeueAfter: time.Second * 5}, r.Status().Update(ctx, externalIP)
-	}
-
-	// 3rd STEP
-	//
-	// Remove finalizer to release ExternalIP
-	if externalIP.Status.State == v1alpha1.ExternalIPStateNone {
-		if controllerutil.ContainsFinalizer(externalIP, externalIPFinalizer) {
-			controllerutil.RemoveFinalizer(externalIP, externalIPFinalizer)
-			return ctrl.Result{}, r.Update(ctx, externalIP)
+	} else {
+		if err := r.Provider.ReconcileExternalIPDeletion(ctx, log, externalIP); err != nil {
+			log.Error(err, "Failed to reconcile ExternalIP deletion")
+			return ctrl.Result{}, err
 		}
 	}
 
-	return ctrl.Result{}, nil
-}
-
-// disassociateAddress performs address disassociation tasks
-func (r *ExternalIPReconciler) disassociateAddress(
-	ctx context.Context,
-	pvd provider.Provider,
-	log logr.Logger,
-	externalIP *v1alpha1.ExternalIP,
-) (ctrl.Result, error) {
-	// Get address and disassociate it
-	if externalIP.Status.AddressID != nil {
-		res, err := pvd.GetAddress(ctx, *externalIP.Status.AddressID)
-		if err != nil {
-			log.Error(err, "Failed to retrieve address", "addressID", *externalIP.Status.AddressID)
+	if !externalIP.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(externalIP, externalIPFinalizer) {
+		controllerutil.RemoveFinalizer(externalIP, externalIPFinalizer)
+		if err := r.Update(ctx, externalIP); err != nil {
+			log.Error(err, "Failed to remove finalizer")
 			return ctrl.Result{}, err
 		}
-
-		if res.AssociationID != nil {
-			if err := pvd.DisassociateAddress(ctx, provider.DisassociateAddressRequest{
-				AssociationID: *res.AssociationID,
-			}); err != nil {
-				log.Error(err, "Failed to disassociate address", "addressID", *externalIP.Status.AddressID, "instanceID", *externalIP.Status.InstanceID)
-				return ctrl.Result{}, err
-			}
-			log.Info("Disassociated address", "addressID", *externalIP.Status.AddressID, "instanceID", *externalIP.Status.InstanceID)
-		}
+		log.V(1).Info("Successfully removed finalizer")
+		return ctrl.Result{}, nil
 	}
```

Comment:
I think this could be simplified like so:
```suggestion
	} else {
		if err := r.Provider.ReconcileExternalIPDeletion(ctx, log, externalIP); err != nil {
			log.Error(err, "Failed to reconcile ExternalIP deletion")
			return ctrl.Result{}, err
		}

        if controllerutil.RemoveFinalizer(externalIP, externalIPFinalizer) {
			if err := r.Update(ctx, externalIP); err != nil {
				log.Error(err, "Failed to remove finalizer")
				return ctrl.Result{}, err
			}
			log.V(1).Info("Successfully removed finalizer")
			return ctrl.Result{}, nil
		}
	}
```

Reply:
Hm no I don't know why my IDE put the suggestion on these lines, I will try to find where I was suggesting this

Reply:
until then